src/zarr/store/zip.py

```diff
@@ -115,6 +117,11 @@ async def empty(self) -> bool:
             else:
                 return True
 
+    def with_mode(self, mode: ZipStoreAccessModeLiteral) -> Self:  # type: ignore[override]
+        return type(self)(
```

Comment:
This probably needs some work... would we want the new Store to share the same Lock?

Reply:
This wasn't straightforward to implement, because of a need to close the old ZipStore (and writing the ending records) before opening the new store.

For now I've opted to raise NotImplementedError for ZipStore.